### PR TITLE
sql+storage: allow ingesting unsupported types as text

### LINKED ISSUE #15716: [Epic] Casting Column Types for Postgres Sources
### Feature request

In https://github.com/MaterializeInc/materialize/issues/14016, we removed `CREATE VIEWS`. This left us with the limitation that if a user has a table in the publication that contains an unsupported column type, the source will crash. We need to add a more fine-grained control as a workaround.

Proposal for the UX (stolen from @sjwiesman because I like it):

```
CREATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
      PUBLICATION 'mz_source',
      CAST foo.my_enum AS string)
FOR ALL TABLES
WITH (SIZE = '3xsmall');
```

We also need to consider the case where the user fails to specify all columns that contain an unsupported type. In this case, I like @benesch's previous suggestion that we provide an error message that tells the user how they can fix their `CREATE SOURCE` statement, e.g.:

```
ERROR: Column "foo" in table "bar" contains an unsupported type. Please use CAST in the CREATE SOURCE statement.
```

### QA Tasks

TBD

### LINKED ISSUE #15889: Postgres Sources May Try to Snapshot Unrelated Table
### What version of Materialize are you using?

Cloud

### How did you install Materialize?

Materialize Cloud

### What is the issue?

A user had a publication with multiple tables, but only specified one in their `FOR TABLES (...)` statement. However, the user was encountered errors of the form `ERROR:  Source error: X: failed during initialization, must be dropped and recreated: missing data for column`. Once they created a publication containing only the one table they wanted to read into Materialize, they no longer saw the error.

This signifies two possible bugs:
1. It looks like we're trying to snapshot an unrelated table.
2. It looks like there is a bug in our `COPY` parser.

Additional context can be found [here](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1667495704567519?thread_ts=1667325066.316519&cid=C01CFKM1QRF).

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by sploiselle**
Redo of #15856

### Motivation

This PR adds a known-desirable feature.
- Closes #15716

This PR fixes a recognized bug.
- Fixes #15889

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Support ingesting unsupported types' data from PostgreSQL sources as text using the new `TEXT COLUMNS` option. For example:

  ```sql
  CREATE SOURCE pg_source
  FROM POSTGRES
  CONNECTION pgconn
  (
    PUBLICATION 'pg_pub',
    TEXT COLUMNS (tbl.col_of_unsupported_type)
  )
  FOR ALL TABLES;
  ```

  This statement will decode the column `tbl.col_of_unsupported_type` into its
  `text` representation. For example, with the PostgreSQL enum type, this ingests
   the actual strings that comprise the enum's values.

### CODE REVIEW

**Comment by sploiselle** [comment]
@petrosagg Derp--just realize I forgot to add wrappers around the oids using `repr::adt::system::Oid`. Will include that in the next push, but would be very glad if I could get the rest of the code reviewed.

File: src/storage-client/src/types/sources.proto [hunk 1]
```diff
@@ -193,18 +193,26 @@ message ProtoKinesisSourceConnection {
...
+        repeated string col_names = 1;
+    }
+
     mz_repr.global_id.ProtoGlobalId connection_id = 6;
     mz_storage_client.types.connections.ProtoPostgresConnection connection = 1;
     string publication = 2;
-    ProtoPostgresSourceDetails details = 4;
-    repeated ProtoPostgresTableCast table_casts = 5;
+    ProtoPostgresSourcePublicationDetails details = 4;
+    map<uint64, ProtoPostgresTableCast> table_casts_map = 7;
```
Comment (petrosagg) [inline src/storage-client/src/types/sources.proto:211]: I haven't made up my mind if this will be a good suggestion but I'm putting it here for elaboration. An alternative that wouldn't break backward compatibility and wouldn't add a reserved field would be to only add an extra field of `repeated uint64` that contains the active/selected tables.

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -735,6 +764,86 @@ pub fn plan_create_source(
...
+            let mut requested_subsources = vec![];
+            for subsource in subsources {
+                let name = subsource.reference.clone();
+
+                let target = match &subsource.subsource {
+                    Some(subsource) => match subsource {
+                        DeferredObjectName::Named(target) => target.clone(),
+                        DeferredObjectName::Deferred(name) => {
+                            // TODO: remove this after the next release, we need
+                            // it only so we can load the catalog to do the
```
Comment (petrosagg) [inline src/sql/src/plan/statement/ddl.rs:779]: Since you're here, can we remove this now? If it's more complicated than just deleting these lines feel free to ignore this comment and doing a separate PR

File: src/sql/src/pure.rs [hunk 3]
```diff
@@ -281,23 +281,74 @@ pub async fn purify_create_source(
...
+                }
+
+                let new = text_cols_dict.entry(oid).or_default().insert(col_name);
+
+                if !new {
+                    bail!("invalid TEXT COLUMNS: multiple references to {name}");
+                }
+            }
+
+            // Normalize options to contain resolved values.
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql/src/pure.rs]: We chatted about this offline; implemented this using a new `ErsatzCatalog` struct.

File: src/sql/src/pure.rs [hunk 4]
```diff
@@ -281,23 +281,74 @@ pub async fn purify_create_source(
...
             let publication = publication
                 .ok_or_else(|| sql_err!("POSTGRES CONNECTION must specify PUBLICATION"))?;
 
             // verify that we can connect upstream and snapshot publication metadata
             let config = connection
                 .config(&*connection_context.secrets_reader)
                 .await?;
-            let tables = mz_postgres_util::publication_info(&config, &publication).await?;
+
+            let mut text_cols_dict: HashMap<u32, HashSet<String>> = HashMap::new();
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql/src/pure.rs]: Because of the way we handle PG subsources, we need to know if a column of an unsupported type should actually be allowed because the user specified it in `TEXT COLUMNS`––this is also where we generate the column definitions for `CREATE SUBSOURCE` statements, so we need to know that it should be treated as text. Not really a simple way to avoid doing this.

File: src/storage/src/source/postgres.rs [hunk 5]
```diff
@@ -1029,10 +1048,24 @@ impl PostgresTaskInfo {
                             // check the name, type_oid, and type_mod explicitly and error if any
                             // of them differ
                             for (src, rel) in info.desc.columns.iter().zip(relation.columns()) {
+                                let mut rel_typoid = u32::try_from(rel.type_id()).unwrap();
+                                let mut rel_typmod = rel.type_modifier();
+
+                                // Rewrite these columns to text if that is how
+                                // we're treating them.
+                                if let Some(text_cols) = self.text_cols.get(&rel_id) {
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/storage/src/source/postgres.rs]: Great catch.

File: src/storage-client/src/types/sources.proto (hunk 1 above)
Comment (sploiselle) [reply src/storage-client/src/types/sources.proto]: Implemented this, as well as an approach for introducing mixed-version compatibility. idk if it's the best approach but I believe it should work.

File: src/sql/src/plan/statement/ddl.rs (hunk 2 above)
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:817]: Sadly the upgrade framework doesn't reflect our current upgrade process (i.e. it takes every version to the current version and does not presume a linear path along versions), which makes this fail. Will file a ticket about this.

**Comment by sploiselle** [comment]
@petrosagg PTAL

File: src/sql/src/plan/statement/ddl.rs [hunk 6]
```diff
@@ -658,8 +696,23 @@ pub fn plan_create_source(
...
+                            match mz_pgrepr::Type::from_oid_and_typmod(
+                                column.type_oid,
+                                column.type_mod,
+                            ) {
+                                Ok(t) => t,
+                                // Skipping this table cast is OK because we
+                                // have validated that this table from the
+                                // publication is not being used during
+                                // purification
+                                Err(_) => continue 'table,
```
Comment (petrosagg) [inline src/sql/src/plan/statement/ddl.rs:711]: Hm, the resulting `column_casts` seems a bit brittle as it doesn't have the same shape for all the tables in the publication. I would suggest always populating the cast expression and when we can't find the type we can just put an expression that calls `mz_error_if_null(NULL, "Unsupported type with OID {column.type_oid}")`

File: src/storage/src/source/postgres.rs [hunk 7]
```diff
@@ -246,28 +248,58 @@ impl SourceConnectionBuilder for PostgresSourceConnection {
...
+                match self.table_casts.get(&output_index) {
+                    Some(casts) => {
+                        let mut desc = desc.clone();
+
+                        // Rewrite all `self.text_cols` columns to text. This
+                        // is the expected output type generated by `casts`
+                        // done in planning, and this must be done here so as
+                        // to not overwrite the "native" values snapshot in
+                        // `self.publication_details`.
+                        if let Some(columns) = self.text_cols.get(&Oid(desc.oid)) {
```
Comment (petrosagg) [inline src/storage/src/source/postgres.rs:269]: Why do we need to alter the description here? I think we should be able to remove the `text_cols` field from `PostgresSourceConnection` entirely because it already contains correctly picked cast expressions selected by the planner

File: src/sql-parser/src/parser.rs [hunk 8]
```diff
@@ -2693,9 +2693,30 @@ impl<'a> Parser<'a> {
...
+        let name = match self.expect_one_of_keywords(&[DETAILS, PUBLICATION, TEXT])? {
             DETAILS => PgConfigOptionName::Details,
             PUBLICATION => PgConfigOptionName::Publication,
+            TEXT => {
+                self.expect_keyword(COLUMNS)?;
+
+                let _ = self.consume_token(&Token::Eq);
+
+                let value = self
+                    .parse_option_sequence(Parser::parse_object_name)?
```
Comment (petrosagg) [inline src/sql-parser/src/parser.rs:2705]: Is this how we parse normal column expressions? Will it work if someone writes `TEXT COLUMNS ("database"."schema"."table"."column")` which contains four elements?

**Review by petrosagg — COMMENTED** [review]
Looks much better!

**Comment by philip-stoev** [comment]
I would like to give this a try first thing tomorrow. Thanks!

File: src/sql/src/plan/statement/ddl.rs (hunk 6 above)
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs]: Sure thing.

File: src/storage/src/source/postgres.rs (hunk 7 above)
Comment (sploiselle) [reply src/storage/src/source/postgres.rs]: I had thought that the column descriptions were used to express how to decode the specific columns from the PG source; hadn't put together that logic is entirely determined by `column_casts`. I'm actually able to remove all of the `text_cols` logic from storage--thanks for pointing this out.

**Comment by sploiselle** [comment]
@petrosagg PTAL

@philip-stoev Sounds good

File: test/pg-cdc/pg-cdc.td [hunk 9]
```diff
@@ -226,9 +262,9 @@ contains: SOURCE materialize.public.conflict_table is a subsource and cannot be
 ! DROP SOURCE conflict_table CASCADE
 contains: SOURCE materialize.public.conflict_table is a subsource and cannot be dropped independently of its primary source
 
-> SELECT * FROM no_replica_identity;
-1
-2
+# > SELECT * FROM no_replica_identity;
```
Comment (philip-stoev) [inline test/pg-cdc/pg-cdc.td:265]: this test is commented out?

File: src/storage-client/src/types/sources.proto [hunk 10]
```diff
@@ -200,11 +200,12 @@ message ProtoPostgresSourceConnection {
     mz_repr.global_id.ProtoGlobalId connection_id = 6;
     mz_storage_client.types.connections.ProtoPostgresConnection connection = 1;
     string publication = 2;
-    ProtoPostgresSourceDetails details = 4;
```
Comment (philip-stoev) [inline src/storage-client/src/types/sources.proto:203]: @aalexandrov has requested that he is notified about any protobuf changes, so I am pinging him here.

File: test/pg-cdc/types-unsupported.td [hunk 11]
```diff
@@ -28,4 +28,4 @@ CREATE PUBLICATION mz_source FOR ALL TABLES;
 
 # Todo: add test for detail and hint once testdrive supports that
 ! CREATE SOURCE mz_source FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source') FOR ALL TABLES;
-contains: column "person.current_mood" uses unrecognized type
+regex: the following columns contain unsupported types:\npostgres\.public\.person\.current_mood \(OID \d+\)
```
Comment (philip-stoev) [inline test/pg-cdc/types-unsupported.td:31]: This is now possible in testdrive, see recently merged PR:

https://github.com/MaterializeInc/materialize/pull/16030/files#diff-0e75185705f0683fc07a43771efee7c26becb34e9b7d5e82db62e29619f9d284

File: src/storage-client/src/types/sources.proto (hunk 10 above)
Comment (aalexandrov) [reply src/storage-client/src/types/sources.proto:203]: This is just renaming the struct, thanks.

File: src/sql/src/plan/statement/ddl.rs [hunk 12]
```diff
@@ -681,18 +764,31 @@ pub fn plan_create_source(
                     );
                     column_casts.push(cast_expr);
                 }
-                table_casts.push(column_casts);
+                let r = table_casts.insert(i + 1, column_casts);
```
Comment (petrosagg) [inline src/sql/src/plan/statement/ddl.rs:767]: Now that the above loop never fails we should be able to remove the `continue` and leave type here as `Vec<Vec<MirScalarExpr>>` since there is exactly one expr for each column for each table, right?

File: src/storage-client/src/types/sources.proto [hunk 13]
```diff
@@ -200,11 +200,12 @@ message ProtoPostgresSourceConnection {
...
+    ProtoPostgresSourcePublicationDetails details = 4;
     repeated ProtoPostgresTableCast table_casts = 5;
+    repeated uint64 table_cast_pos = 7;
```
Comment (petrosagg) [inline src/storage-client/src/types/sources.proto:205]: Is this representing the selected tables? The same seems a bit hard to parse. It's the position of table casts?

File: src/sql/src/plan/statement/ddl.rs (hunk 12 above)
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:767]: Certainly possible, but requires more thoughtful accounting when retaining the requested table casts here. Glad to make that change if you prefer; was just trying to avoid as many off by one errors as possible.

```rust
    if let GenericSourceConnection::Postgres(PostgresSourceConnection { table_casts, .. }) =
        &mut external_connection
    {
        // Now that we know which subsources sources we want, we can remove all
        // unused table casts from this connection; this represents the
        // authoritative statement about which publication tables should be
        // used within storage.
        let used_pos: HashSet<_> = subsource_exports.values().collect();
        table_casts.retain(|pos, _| used_pos.contains(pos));
    }
 ```

File: src/storage-client/src/types/sources.proto (hunk 13 above)
Comment (sploiselle) [reply src/storage-client/src/types/sources.proto:205]: It's the position in the PG source to which the table cast correlates. This is called the "output index" elsewhere. If you have a suggestion for an improved name, glad to adopt it, or can be more explicit in the code about the field's intended semantics.

File: test/pg-cdc/types-unsupported.td (hunk 11 above)
Comment (philip-stoev) [reply test/pg-cdc/types-unsupported.td:31]: I mentioned this because of this comment: `Todo: add test for detail and hint once testdrive supports that`

Comment (sploiselle) [reply test/pg-cdc/types-unsupported.td:31]: Oh sorry--I totally overlooked the `Todo`; ty

File: src/sql/src/plan/statement/ddl.rs [hunk 14]
```diff
@@ -658,8 +699,50 @@ pub fn plan_create_source(
...
+                                                        .as_str(),
+                                                    ),
+                                                    ScalarType::String,
+                                                ),
+                                            ],
+                                        }
+                                        .lower_uncorrelated()
+                                        .expect("no correlation"),
+                                    );
+                                    continue;
```
Comment (petrosagg) [inline src/sql/src/plan/statement/ddl.rs:740]:
```suggestion
```

Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:740]: This continue is required because we are not returning an `mz_pgrepr::Type::Text` from this match arm

**Review by petrosagg — APPROVED** [review]
My pending comments can be fixed in post, the PR overall looks great and the feature works so let's ship it!

**Comment by umanwizard** [comment]
For posterity: force merged this without waiting for CI, because it's a release blocker, and we are going to do a full CI run on it for the release anyway.
